include/zephyr/irq.h

```diff
@@ -266,22 +272,25 @@ void z_smp_global_unlock(unsigned int key);
 static inline unsigned int irq_get_level(unsigned int irq)
 {
 #if defined(CONFIG_3RD_LEVEL_INTERRUPTS)
-	return ((irq >> 16) & 0xFF) != 0 ? 3 :
-		(((irq >> 8) & 0xFF) == 0 ? 1 : 2);
+	return ((irq >> (CONFIG_1ST_LEVEL_INTERRUPT_BITS + CONFIG_2ND_LEVEL_INTERRUPT_BITS))
+		& BIT_MASK(CONFIG_3RD_LEVEL_INTERRUPT_BITS)) != 0 ?
+		3 : (((irq >> CONFIG_2ND_LEVEL_INTERRUPT_BITS)
+			& BIT_MASK(CONFIG_2ND_LEVEL_INTERRUPT_BITS)) == 0 ? 1 : 2);
 #elif defined(CONFIG_2ND_LEVEL_INTERRUPTS)
-	return ((irq >> 8) & 0xFF) == 0 ? 1 : 2;
+	return ((irq >> CONFIG_2ND_LEVEL_INTERRUPT_BITS) &
+		BIT_MASK(CONFIG_2ND_LEVEL_INTERRUPT_BITS)) == 0 ?
+		1 : 2;
 #else
 	ARG_UNUSED(irq);
 
 	return 1;
 #endif
```

Comment:
This should also do the trick and is still optimizable by the compiler.
https://godbolt.org/z/WMW81xrch

```cpp
 const uint32_t mask2 = BIT_MASK(CONFIG_2ND_LEVEL_INTERRUPT_BITS) << CONFIG_1ST_LEVEL_INTERRUPT_BITS;
 const uint32_t mask3 = BIT_MASK(CONFIG_3RD_LEVEL_INTERRUPT_BITS) << CONFIG_1ST_LEVEL_INTERRUPT_BITS + CONFIG_2ND_LEVEL_INTERRUPT_BITS;

 if (IS_ENABLED(CONFIG_3RD_LEVEL_INTERRUPTS) && (irq & mask3) != 0) {
   return 3;
 }

 if (IS_ENABLED(CONFIG_2ND_LEVEL_INTERRUPTS) && (irq & mask2) != 0) {
   return 2;
 }

 return 1;
```